tests/syscall/epoll/enc/enc.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <netinet/in.h>
+#include <openenclave/enclave.h>
+#include <openenclave/internal/tests.h>
+#include <sys/epoll.h>
+#include <sys/socket.h>
+#include <unistd.h>
+#include <cerrno>
+#include <cstdio>
+
+static const uint16_t _port = 12347;
+static sockaddr_in _addr;
+static int _epfd;
+static int _sockfd;
+
+extern "C" void set_up()
+{
+    OE_TEST(oe_load_module_host_socket_interface() == OE_OK);
+    OE_TEST(oe_load_module_host_epoll() == OE_OK);
+
+    _addr.sin_family = AF_INET;
+    _addr.sin_port = htons(_port);
+    _addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+
+    // create and bind UDP socket
+    _sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+    OE_TEST(_sockfd >= 0);
+    OE_TEST(
+        bind(_sockfd, reinterpret_cast<sockaddr*>(&_addr), sizeof(_addr)) == 0);
+
+    // create epoll instance
+    _epfd = epoll_create(1);
+    OE_TEST(_epfd >= 0);
+}
+
+extern "C" void tear_down()
+{
+    OE_TEST(close(_epfd) == 0);
+    OE_TEST(close(_sockfd) == 0);
+}
+
+extern "C" void wait_for_events()
+{
+    uint8_t run = 1;
+
+    while (run)
+    {
+        epoll_event event{};
+
+        int n;
+        do
+        {
+            n = epoll_wait(_epfd, &event, 1, -1);
+        } while (n == -1 && errno == EINTR);
+
+        if (n == 1)
+        {
+            OE_TEST(event.data.fd == _sockfd);
+            OE_TEST(read(_sockfd, &run, sizeof(run)) == sizeof(run));
+        }
+        else
+            OE_TEST(n == 0); // fd has been deleted
+    }
+}
+
+static void _send(uint8_t run)
+{
+    const int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+    OE_TEST(sockfd >= 0);
+    OE_TEST(
+        connect(sockfd, reinterpret_cast<sockaddr*>(&_addr), sizeof(_addr)) ==
+        0);
+    OE_TEST(write(sockfd, &run, sizeof(run)) == sizeof(run));
+    OE_TEST(close(sockfd) == 0);
+}
+
+extern "C" void trigger_and_add_event()
+{
+    _send(1);
+
+    // add fd to the epoll instance
+    epoll_event event{};
+    event.events = EPOLLIN;
+    event.data.fd = _sockfd;
+    OE_TEST(epoll_ctl(_epfd, EPOLL_CTL_ADD, _sockfd, &event) == 0);
+}
+
+extern "C" void trigger_and_delete_event()
+{
+    _send(1);
+
+    // delete fd from the epoll instance
+    OE_TEST(epoll_ctl(_epfd, EPOLL_CTL_DEL, _sockfd, nullptr) == 0);
+}
+
+extern "C" void cancel_wait()
```

Comment:
Why is this called `cancel_wait`? That doesn't look like what the function is doing.

Reply:
How about `send_stop_request`?

Reply:
Sorry I should have been more clear. I don't understand how the below code would cancel the thread from waiting. Doesn't it just add a wait on a handle and then trigger it? Why would epoll stop waiting? `wait_for_events` has a `while (1)` loop so I don't see why it would ever stop unless one of the OE_TEST calls fails.

Excuse my likely ignorance of epoll

Reply:
The code was a little cryptic, I admit. Improved.